falcon/asgi/request.py

```diff
@@ -794,26 +802,42 @@ def if_match(self):
         #   gets a None back on the first reference to property, it
         #   probably isn't going to access the property again (TBD).
         if self._cached_if_match is None:
-            header_value = self._asgi_headers.get('if-match')
+            header_value = self._asgi_headers.get(b'if-match')
             if header_value:
-                self._cached_if_match = helpers._parse_etags(header_value)
+                self._cached_if_match = helpers._parse_etags(header_value.decode())
 
         return self._cached_if_match
 
     @property
     def if_none_match(self):
         if self._cached_if_none_match is None:
-            header_value = self._asgi_headers.get('if-none-match')
+            header_value = self._asgi_headers.get(b'if-none-match')
             if header_value:
-                self._cached_if_none_match = helpers._parse_etags(header_value)
+                self._cached_if_none_match = helpers._parse_etags(header_value.decode())
 
         return self._cached_if_none_match
 
+    @property
+    def headers(self):
+        # NOTE(kgriffs: First time here will cache the dict so all we
+        # have to do is clone it in the future.
+        if self._cached_headers is None:
+            self._cached_headers = {
+                name.decode(): value.decode()
+                for name, value in self._asgi_headers.items()
+            }
+
+        return self._cached_headers
```

Comment:
As much as I am a huge fan of comprehensions, a simple loop seems to outperform them more often than not.

On 3.8:
```python
>>> headers = {b'content-type': b'application/yaml', b'x-coolness': b'moderate', b'user-agent': b'Browser', b'via': b'SomeCDN', b'content-length': b'8', b'x-some': b'thing'}
>>> def func1():
...     return {key.decode(): value.decode() for key, value in headers.items()}
... 
>>> func1()
{'content-type': 'application/yaml', 'x-coolness': 'moderate', 'user-agent': 'Browser', 'via': 'SomeCDN', 'content-length': '8', 'x-some': 'thing'}
>>> def func2():
...     result = {}
...     for key, value in headers.items():
...         result[key.decode()] = value.decode()
...     return result
... 
>>> func2()
{'content-type': 'application/yaml', 'x-coolness': 'moderate', 'user-agent': 'Browser', 'via': 'SomeCDN', 'content-length': '8', 'x-some': 'thing'}
>>> timeit.timeit(func1)
1.2187647159989865
>>> timeit.timeit(func1)
1.216893122997135
>>> timeit.timeit(func1)
1.21874491300332
>>> timeit.timeit(func2)
1.1081436029999168
>>> timeit.timeit(func2)
1.1094474159981473
>>> timeit.timeit(func2)
1.116244266002468
```